File: Templates/TypeEditor/src/Templates/Templates/CS.TypeEditor/CustomControlLibrary.Protocol/Endpoints/CreateCustomTypeEditorVMResponse.cs
using System;
using System.Diagnostics.CodeAnalysis;
using CustomControlLibrary.Protocol.DataTransport;
using Microsoft.DotNet.DesignTools.Protocol.DataPipe;
using Microsoft.DotNet.DesignTools.Protocol.Endpoints;

namespace CustomControlLibrary.Protocol.Endpoints
{
    /// <summary>
    ///  Response class, answering the request for that endpoint. This transports the requested data (Proxy of
    ///  the server-side ViewModel and the data of the custom property type <c>PropertyStore</c>) back to the client.
    /// </summary>
    public class CreateCustomTypeEditorVMResponse : Response
    {
        [AllowNull]
        public object ViewModel { get; private set; }

        public CustomPropertyStoreData? PropertyStoreData { get; set; }

        public CreateCustomTypeEditorVMResponse() { }

        public CreateCustomTypeEditorVMResponse(
            object viewModel,
            CustomPropertyStoreData? propertyStoreData)
        {
            ViewModel = viewModel ?? throw new ArgumentNullException(nameof(viewModel));
            PropertyStoreData = propertyStoreData;
        }

        public CreateCustomTypeEditorVMResponse(IDataPipeReader reader) : base(reader) { }

        protected override void ReadProperties(IDataPipeReader reader)
        {
            ViewModel = reader.ReadObject(nameof(ViewModel));
            PropertyStoreData = reader.ReadDataPipeObjectOrNull<CustomPropertyStoreData>(nameof(PropertyStoreData));
        }

        protected override void WriteProperties(IDataPipeWriter writer)
        {
            writer.WriteObject(nameof(ViewModel), ViewModel);
            writer.WriteDataPipeObjectIfNotNull<CustomPropertyStoreData>(nameof(PropertyStoreData), PropertyStoreData!);
        }
    }
}

File: Templates/TypeEditor/src/Templates/Templates/VB.TypeEditor/CustomControlLibrary.Protocol/Endpoints/CreateCustomTypeEditorVMResponse.vb
Imports CustomControlLibrary.Protocol.DataTransport
Imports Microsoft.DotNet.DesignTools.Protocol.DataPipe
Imports Microsoft.DotNet.DesignTools.Protocol.Endpoints

Namespace Endpoints
    ''' <summary>
    '''  Response class, answering the request for that endpoint. This transports the requested data (Proxy of
    '''  the server-side ViewModel and the data of the custom property type <c>PropertyStore</c>) back to the client.
    ''' </summary>
    Public Class CreateCustomTypeEditorVMResponse
        Inherits Response

        Private privateViewModel As Object

        Public Property ViewModel() As Object
            Get
                Return privateViewModel
            End Get
            Private Set(ByVal value As Object)
                privateViewModel = value
            End Set
        End Property

        Public Property PropertyStoreData() As CustomPropertyStoreData

        Public Sub New()
        End Sub

        Public Sub New(viewModel As Object, propertyStoreData As CustomPropertyStoreData)
            Me.ViewModel = GlobalUtilities.OrThrowIfArgumentIsNull(viewModel)
            Me.PropertyStoreData = propertyStoreData
        End Sub

        Public Sub New(ByVal reader As IDataPipeReader)
            MyBase.New(reader)
        End Sub

        Protected Overrides Sub ReadProperties(ByVal reader As IDataPipeReader)
            ViewModel = reader.ReadObject(NameOf(ViewModel))
            PropertyStoreData = reader.ReadDataPipeObjectOrNull(Of CustomPropertyStoreData)(NameOf(PropertyStoreData))
        End Sub

        Protected Overrides Sub WriteProperties(ByVal writer As IDataPipeWriter)
            'Should be like this, but the compiler disagrees, so we use a workaround.
            DataPipeWriterExtensions.WriteObject(writer, NameOf(ViewModel), ViewModel)
            writer.WriteDataPipeObjectIfNotNull(Of CustomPropertyStoreData)(NameOf(PropertyStoreData), PropertyStoreData)
        End Sub
    End Class
End Namespace
